modules/gapi/include/opencv2/gapi/gtransform.hpp

```diff
@@ -0,0 +1,210 @@
+// consider better approach like in compound kernels with tuple wrappers and context class
+
+template <typename K, typename... Ins, typename Out>
+struct TransHelper<K, std::tuple<Ins...>, Out>
+{
+    // FIXME: code duplication
+    template <int... IIs>
+    static GArgs get_pattern_impl(const GArgs &in_args, detail::Seq<IIs...>)
+    {
+        using R = typename TransType<Out>::type;
+        const R r = K::pattern(get_in_args<Ins>(in_args, IIs)...);
+        return GArgs{GArg(r)};
+    }
+
+    template <int... IIs>
+    static GArgs get_substitute_impl(const GArgs &in_args, detail::Seq<IIs...>)
+    {
+        using R = typename TransType<Out>::type;
+        const R r = K::substitute(get_in_args<Ins>(in_args, IIs)...);
+        return GArgs{GArg(r)};
+    }
+
+    static GArgs get_pattern(const GArgs &in_args)
+    {
+        return get_pattern_impl(in_args, typename detail::MkSeq<sizeof...(Ins)>::type());
+    }
+    static GArgs get_substitute(const GArgs &in_args)
+    {
+        return get_substitute_impl(in_args, typename detail::MkSeq<sizeof...(Ins)>::type());
+    }
+};
+
+template <typename K, typename... Ins, typename... Outs>
+struct TransHelper<K, std::tuple<Ins...>, std::tuple<Outs...>>
+{
+    // FIXME: code duplication
+    template <int... IIs, int... OIs>
+    static GArgs get_pattern_impl(const GArgs &in_args, detail::Seq<IIs...>, detail::Seq<OIs...>)
+    {
+        using R = std::tuple<typename TransType<Outs>::type...>;
+        const R r = K::pattern(get_in_args<Ins>(in_args, IIs)...);
+        return GArgs{GArg(std::get<OIs>(r))...};
+    }
+
+    template <int... IIs, int... OIs>
+    static GArgs get_substitute_impl(const GArgs &in_args, detail::Seq<IIs...>, detail::Seq<OIs...>)
+    {
+        using R = std::tuple<typename TransType<Outs>::type...>;
+        const R r = K::substitute(get_in_args<Ins>(in_args, IIs)...);
+        return GArgs{GArg(std::get<OIs>(r))...};
+    }
+
+    static GArgs get_pattern(const GArgs &in_args)
+    {
+        return get_pattern_impl(in_args, typename detail::MkSeq<sizeof...(Ins)>::type(),
+                                typename detail::MkSeq<sizeof...(Outs)>::type());
+    }
+    static GArgs get_substitute(const GArgs &in_args)
+    {
+        return get_substitute_impl(in_args, typename detail::MkSeq<sizeof...(Ins)>::type(),
+                                   typename detail::MkSeq<sizeof...(Outs)>::type());
+    }
+};
+
+/////////////////////////////////////////////////////////////////////
+
+template <typename, typename>
+class GTransformImpl;
+
+template <typename K, typename R, typename... Args>
+class GTransformImpl<K, std::function<R(Args...)>> : public TransHelper<K, std::tuple<Args...>, R>,
+                                                     public cv::detail::TransformTag
+{
+public:
+    // FIXME: currently there is no check that transformations' signatures are unique
+    // and won't be any intersection in graph compilation stage
+    using API = K;
```

Comment:
This is kinda workaround. Need to think about 2 signatures having same signature but different kernels in use.

Reply:
Why do you need `API` defined here BTW? Kernel package (which used that) should be already neutral to this, I believe.

Reply:
`API` is checked on uniqueness via `all_unique` in kernels() method

Reply:
Oh, this is nice, thanks!